ubiquity/ubiquity-dollar
# feat: show alert on phishing attempt

Resolves #343 

Yes, trying to merge obfuscated code into the public crypto project :)

Simple thing is that you can not make user safe in case of DNS hijacking as the whole website is controlled by the attacker.

What we can do is to warn user that something is wrong and make him to report it ASAP.

Typically the attacker goes this way:
1. Copies website frontend
2. Resolves all path errors
3. Adds malicious scripts

I tried it all myself and resolving all next js production build paths (js, css, images, etc...) is a tedious task but feasible.  

We can not put our "warn script" in an external script as the attacker will probably just delete it (as I did going the "attacker way"), we can not use WASM as it doesn't have access to the DOM (but yes, disassembling is a hard task), we can not hide it inside svg or meta tags as next js production build just deletes anything what it feels to be malicious. That is why our "warn script" should reside in `index.html`(in which way this script will not produce any unresolved path errors) as there are chances that the attacker will not delete it resolving path errors.

When domain is hijacked or website is just copied with malicious attempts we should somehow separate "real website" and "malicious website". Nextjs is a node js framework meaning there is a server side behind the website. When attacker copies website he just copies the frontend part without the backend. So if we know that we have a particular API on our backend, call a particular API on website load and it responds with 200 then we can be pretty much sure that we are on a "real website" as "malicious website" may respond with 404.

So this is the unobfuscated script that is proposed to merge:
```
if (!['localhost','127.0.0.1'].includes(window.location.hostname)) {
    fetch('/api/heartbeat')
      .then((response) => {
        if (!response.ok) {
          document.body.innerHTML = '<div style="background-color: red; padding: 20px;"><span>Phishing attempt! Please contact us at <a style="color: blue" href="https://t.me/ubiquitydao" target="_blank">https://t.me/ubiquitydao</a></span></div>';
        }
    });
}
```

How it works:
1. Check that we are on a real domain (as the attacker will probably debug on a localhost)
2. Call `/api/heartbeat`. Real website will respond with 200. Copied website will respond with 404.
3. If response is not 200 then show error

This is how the error looks like:
<img width="1440" alt="Screenshot 2022-12-27 at 17 38 10" src="https://user-images.githubusercontent.com/119500907/209691892-51e48448-9ff1-4cf1-99c3-042042133fa4.png">

So the script deletes the whole UI and shows only the warning message.

How to test:
1. Copy website frontend (you can use "Save As" in google chrome)
2. Run the copied website via `npx http-server` (it should be served at `localhost:8080`)
3. Add this line to `/etc/hosts`: `127.0.0.1    my.site`
4. Open localhost:8080 (you should see the normal UI as the script doesn't work on localhost)
5. Open `my.site:8080` (you should see the expected warning message)

The code is obfuscated via https://obfuscator.io/ , just paste the js code, check the "Unicode Escape Sequence" option and hit "Obfuscate".
